// @storybook/ui/dist/modules/ui/containers/header.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mapper = undefined;

var _lodash = require('lodash.pick');

var _lodash2 = _interopRequireDefault(_lodash);

var _header = require('../components/header');

var _header2 = _interopRequireDefault(_header);

var _gen_podda_loader = require('../libs/gen_podda_loader');

var _gen_podda_loader2 = _interopRequireDefault(_gen_podda_loader);

var _compose = require('../../../compose');

var _compose2 = _interopRequireDefault(_compose);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var mapper = exports.mapper = function mapper(state, props, _ref) {
  var actions = _ref.actions;

  var currentOptions = (0, _lodash2.default)(state.shortcutOptions, 'showStoriesPanel', 'enableShortcuts', 'addonPanelInRight');

  var actionMap = actions();
  var uiOptions = state.uiOptions,
      isMobileDevice = state.isMobileDevice;
  var _uiOptions$name = uiOptions.name,
      name = _uiOptions$name === undefined ? '' : _uiOptions$name,
      _uiOptions$url = uiOptions.url,
      url = _uiOptions$url === undefined ? '' : _uiOptions$url;


  var handleBurgerButtonClick = function handleBurgerButtonClick() {
    actionMap.shortcuts.setOptions({
      showStoriesPanel: !currentOptions.showStoriesPanel
    });
  };

  var addonPanelInRight = isMobileDevice ? false : currentOptions.addonPanelInRight;

  return {
    name: name,
    url: url,
    enableShortcutsHelp: currentOptions.enableShortcuts,
    openShortcutsHelp: actionMap.ui.toggleShortcutsHelp,
    onBurgerButtonClick: handleBurgerButtonClick,
    isMobileDevice: isMobileDevice,
    addonPanelInRight: addonPanelInRight
  };
};

exports.default = (0, _compose2.default)((0, _gen_podda_loader2.default)(mapper))(_header2.default);
