// @storybook/react/dist/server/config/webpack.config.prod.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (configDir) {
  var entries = {
    preview: [require.resolve('./polyfills'), require.resolve('./globals')],
    manager: [require.resolve('./polyfills'), _server.managerPath]
  };

  var config = {
    mode: 'production',
    bail: true,
    devtool: '#cheap-module-source-map',
    entry: entries,
    output: {
      filename: 'static/[name].[chunkhash].bundle.js',
      // Here we set the publicPath to ''.
      // This allows us to deploy storybook into subpaths like GitHub pages.
      // This works with css and image loaders too.
      // This is working for storybook since, we don't use pushState urls and
      // relative URLs works always.
      publicPath: ''
    },
    plugins: [new _htmlWebpackPlugin2.default({
      filename: 'index.html',
      chunks: ['manager', 'runtime~manager'],
      chunksSortMode: 'none',
      data: {
        managerHead: (0, _server.getManagerHeadHtml)(configDir),
        version: _package.version
      },
      template: _server.indexHtmlPath
    }), new _htmlWebpackPlugin2.default({
      filename: 'iframe.html',
      excludeChunks: ['manager', 'runtime~manager'],
      chunksSortMode: 'none',
      data: {
        previewHead: (0, _server.getPreviewHeadHtml)(configDir)
      },
      template: _server.iframeHtmlPath
    }), new _InterpolateHtmlPlugin2.default(process.env), new _webpack2.default.DefinePlugin((0, _utils.loadEnv)({ production: true })), new _dotenvWebpack2.default({ silent: true })],
    module: {
      rules: [{
        test: /\.jsx?$/,
        loader: require.resolve('babel-loader'),
        query: _babel2.default,
        include: _utils.includePaths,
        exclude: _utils.excludePaths
      }, {
        test: /\.md$/,
        use: [{
          loader: require.resolve('html-loader')
        }, {
          loader: require.resolve('markdown-loader')
        }]
      }]
    },
    resolve: {
      // Since we ship with json-loader always, it's better to move extensions to here
      // from the default config.
      extensions: ['.js', '.json', '.jsx'],
      // Add support to NODE_PATH. With this we could avoid relative path imports.
      // Based on this CRA feature: https://github.com/facebookincubator/create-react-app/issues/253
      modules: ['node_modules'].concat(_utils.nodePaths)
    },
    optimization: {
      // Automatically split vendor and commons for preview bundle
      // https://twitter.com/wSokra/status/969633336732905474
      splitChunks: {
        chunks: function chunks(chunk) {
          return chunk.name !== 'manager';
        }
      },
      // Keep the runtime chunk seperated to enable long term caching
      // https://twitter.com/wSokra/status/969679223278505985
      runtimeChunk: true
    }
  };

  return config;
};

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _dotenvWebpack = require('dotenv-webpack');

var _dotenvWebpack2 = _interopRequireDefault(_dotenvWebpack);

var _InterpolateHtmlPlugin = require('@storybook/react-dev-utils/InterpolateHtmlPlugin');

var _InterpolateHtmlPlugin2 = _interopRequireDefault(_InterpolateHtmlPlugin);

var _htmlWebpackPlugin = require('html-webpack-plugin');

var _htmlWebpackPlugin2 = _interopRequireDefault(_htmlWebpackPlugin);

var _server = require('@storybook/core/server');

var _babel = require('./babel.prod');

var _babel2 = _interopRequireDefault(_babel);

var _utils = require('./utils');

var _package = require('../../../package.json');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
